app/src/main/java/org/dharmaseed/android/PlaybackService.java

```diff
@@ -0,0 +1,232 @@
+import android.util.Log;
+
+import androidx.annotation.Nullable;
+import androidx.annotation.OptIn;
+import androidx.media3.common.AudioAttributes;
+import androidx.media3.common.C;
+import androidx.media3.common.MediaItem;
+import androidx.media3.common.MediaMetadata;
+import androidx.media3.common.Player;
+import androidx.media3.common.util.UnstableApi;
+import androidx.media3.datasource.DefaultDataSource;
+import androidx.media3.datasource.DefaultHttpDataSource;
+import androidx.media3.exoplayer.ExoPlayer;
+import androidx.media3.exoplayer.source.DefaultMediaSourceFactory;
+import androidx.media3.session.CommandButton;
+import androidx.media3.session.DefaultMediaNotificationProvider;
+import androidx.media3.session.MediaNotification;
+import androidx.media3.session.MediaSession;
+import androidx.media3.session.MediaSessionService;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.ListenableFuture;
+
+import java.io.ByteArrayOutputStream;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.text.SimpleDateFormat;
+import java.util.GregorianCalendar;
+import java.util.List;
+import java.util.Timer;
+import java.util.Vector;
+
+public class PlaybackService extends MediaSessionService {
+
+    private static final String LOG_TAG = "PlaybackService";
+    private MediaSession mediaSession;
+    private Runnable updateTalkProgressTask;
+    private Handler handler;
+
+    @Override
+    @OptIn(markerClass = UnstableApi.class)
+    public void onCreate() {
+        super.onCreate();
+
+        // Create the media player
+        Context context = this;
+        DefaultHttpDataSource.Factory httpFactory = new DefaultHttpDataSource.Factory()
+                .setAllowCrossProtocolRedirects(true);
+        DefaultDataSource.Factory dataSourceFactory = new DefaultDataSource.Factory(
+                context, httpFactory
+        );
+
+        AudioAttributes audioAttributes = new AudioAttributes.Builder()
+                .setUsage(C.USAGE_MEDIA)
+                .build();
+
+        ExoPlayer mediaPlayer = new ExoPlayer.Builder(context)
+                .setMediaSourceFactory(
+                        new DefaultMediaSourceFactory(context)
+                                .setDataSourceFactory(dataSourceFactory))
+                .setWakeMode(WAKE_MODE_NETWORK)
+                .setHandleAudioBecomingNoisy(true)
+                .setUseLazyPreparation(true)
+                .setAudioAttributes(audioAttributes, true)
+                .build();
+
+        // Create a media session
+        mediaSession = new MediaSession.Builder(this, mediaPlayer)
+                .setCallback(new SessionCallback())
+                .build();
+
+        // Set a custom media notification
+        setMediaNotificationProvider(new NotificationProvider(this));
+
+        // Set a timer to periodically update the talk progress in the database
+        handler = new Handler();
```

Comment:
@intermarc, I tried to use `Timer` here like you did in `PlayTalkActivity`, but I ran into trouble because as far as I could see, `runOnUiThread` isn't available in this context (a `Service`, not an `Activity`). I think that's required because `ExoPlayer` throws an exception if you call its methods from a thread different from the UI thread. So I went back to using a `Handler` here that stops calling its task once this service is destroyed. I think that works OK but if you have any ideas for how to approach this in a better way, I'm definitely open to them. :)

Reply:
https://github.com/dharmaseed/dharmaseed-android/blob/cd529094dce3acc19dd3bd2012a1843ef1271f44/app/src/main/java/org/dharmaseed/android/PlaybackService.java#L113-L115

I think that this is a very good solution, since the `Hander` stops itself when the `MediaSession` has ended.